sdks/python/apache_beam/runners/portability/portable_runner.py

```diff
@@ -189,20 +189,60 @@ def run_pipeline(self, pipeline, options):
       proto_pipeline = fn_api_runner_transforms.with_stages(
           proto_pipeline, stages)
 
-    # TODO: Define URNs for options.
-    # convert int values: https://issues.apache.org/jira/browse/BEAM-5509
-    p_options = {'beam:option:' + k + ':v1': (str(v) if type(v) == int else v)
-                 for k, v in options.get_all_options().items()
-                 if v is not None}
-
     if not job_service:
       channel = grpc.insecure_channel(job_endpoint)
       grpc.channel_ready_future(channel).result()
       job_service = beam_job_api_pb2_grpc.JobServiceStub(channel)
     else:
       channel = None
 
+    # fetch runner options from job service
+    def send_options_request(max_retries=5):
+      num_retries = 0
+      while True:
+        try:
+          # This reports channel is READY but connections may fail
+          # Seems to be only an issue on Mac with port forwardings
+          if channel:
+            grpc.channel_ready_future(channel).result()
+          return job_service.DescribePipelineOptions(
+              beam_job_api_pb2.DescribePipelineOptionsRequest())
+        except grpc._channel._Rendezvous as e:
+          num_retries += 1
+          if num_retries > max_retries:
+            raise e
+
+    options_response = send_options_request()
+
+    def add_runner_options(parser):
+      for option in options_response.options:
+        try:
+          # no default values - we don't want runner options
+          # added unless they were specified by the user
+          # TODO: types
+          action = 'store'
+          if option.type == 'Boolean':
+            action = 'store_true'
+          parser.add_argument("--%s" % option.name,
+                              action=action,
+                              help=option.description
+                             )
+        except Exception as e:
+          # ignore runner options that are already present
+          # only in this case is duplicate not treated as error
+          if 'conflicting option string' not in str(e):
+            raise
```

Comment:
Could we send the existing options alongside with the DescribeRequest? We could then only send the missing ones.

Reply:
I don't see a way to retrieve existing keys. If that was the case, we could skip the call to add_argument (no need to send them to job server).

Reply:
It would be good to get rid of the `except` check here. This shouldn't be necessary if we provide all runner options via the describe message. We want to raise an exception if the options are duplicated in the SDK.

Reply:
No longer necessary for the Flink options. But there are some other options like `streaming` that still overlap.

Reply:
So it would be possible to send those existing options with the DescribeRequest and filter them out on the Runner side? That way we wouldn't have duplicates.

Reply:
It would require the option parsing to be repeated, first to get the options w/o runner options (where unknown runner options are ignored), then send those to the job service, then repeat parsing with runner options present. I think the return of that isn't worth the extra work and I would like to leave it out of this PR.

Reply:
Could we add a name space prefix to options that come from runners?

Reply:
IMO, from user perspective, these options should work like any other option that for example the SDK provides. Note that "runner options" can belong to multiple groups also. We discussed last week how collision in the global key namespace could be handled by using the option group as explicit qualifier. When there is ambiguity, the user could resolve it by specifying a namespace prefix. For example, --MyOptions.foo and --FlinkOptions.foo  could coexist. This would be separate work, but I can create a JIRA for it. The implementation as of this PR does not represent any change compared to current behavior WRT uniqueness and precedence of option keys.

Reply:
Filed as https://issues.apache.org/jira/browse/BEAM-6531

Reply:
FYI the options that overlap as logged:
```
INFO:root:Runner option 'job_name' was already added
INFO:root:Runner option 'runner' was already added
INFO:root:Runner option 'temp_location' was already added
INFO:root:Runner option 'streaming' was already added
INFO:root:Runner option 'project' was already added
INFO:root:Runner option 'zone' was already added
```
This is not a scoping issue, these are the same options defined in 2 places. There are multiple ways to eliminate this overlap to avoid the exception catch, but I would like to defer this to follow-up.

Reply:
>This is not a scoping issue, these are the same options defined in 2 places.

If they were scoped, they wouldn't be defined twice :) I agree though that this is an old issue which can be addressed separately.

Reply:
Scoping would be useful to allow for different options with the same name, which requires extensive changes across the board. Above log entries show options that are logically the same and need to be recognized by the runner as such, but they are defined in 2 different places..

Reply:
Filing a JIRA and working on this later sounds good.

Reply:
Done: https://issues.apache.org/jira/browse/BEAM-6537